warehouse/utils/admin_flags.py

```diff
@@ -0,0 +1,33 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from sqlalchemy import Column, Boolean, Text
+
+from warehouse import db
+
+
+class AdminFlag(db.Model):
+
+    __tablename__ = "warehouse_admin_flag"
+
+    id = Column(Text, primary_key=True, nullable=False)
+    description = Column(Text, nullable=False)
+    enabled = Column(Boolean)
+
+    @classmethod
+    def is_enabled(cls, session, flag_name):
+        flag = (session.query(AdminFlag)
+                       .filter(AdminFlag.id == flag_name)
+                       .first())
+        if flag is None:
```

Comment:
I'm torn on this, on one hand it's nice that by surviving a flag not existing in the database, that we're able to handle things like the code running prior to migrations having being ran. However I think that generally code can require a migration being ran-- but migrations have to be compatible with the previous iteration of Warehouse right?

In that case I think I would suggest making it a hard error to not have the flag exist in the database, and prepopulate the database in the migration with any new admin flags. That also leads the way to an eventual admin interface for these, where the admin interface itself doesn't allow creating new admin flags (since they require code support, and getting the ID correct is important) but does allow toggling the flag on/off.It also means there is a single place we can see all of the admin flags (in the DB).

Reply:
I'd certainly rather failsafe than make it a hard error if a flag isn't available. Consider the case where a release is being canaried and has _optional_ migrations or works around those in some fashion.

But, I did add the flags we're using into the migration, as indeed it significantly reduces the complexity the Admin interface implementation.